### PR TITLE
feat(schema validation): implement backup restore

### PR DESCRIPTION
**Pull request by thalesmg**
Fixes https://emqx.atlassian.net/browse/EMQX-12346

Release version: e5.7

## Summary

From discussions with Product, when importing a schema validation backup:
 
* Existing validations (identified by `name') are left untouched.
* No validations are removed.
* New validations are appended to the existing list.
* Existing validations are not reordered.

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [x] Added tests for the changes
- [ ] Added property-based tests for code which performs user input validation
- [ ] Changed lines covered in coverage report
- [ ] Change log has been added to `changes/(ce|ee)/(feat|perf|fix|breaking)-<PR-id>.en.md` files
- [x] For internal contributor: there is a jira ticket to track this change
- [ ] Created PR to [emqx-docs](https://github.com/emqx/emqx-docs) if documentation update is required, or link to a follow-up jira ticket
- [ ] Schema changes are backward compatible

## Checklist for CI (.github/workflows) changes

- [ ] If changed package build workflow, pass [this action](https://github.com/emqx/emqx/actions/workflows/build_packages.yaml) (manual trigger)
- [ ] Change log has been added to `changes/` dir for user-facing artifacts update

### CODE REVIEW

File: apps/emqx_schema_validation/src/emqx_schema_validation.erl [hunk 1]
```diff
@@ -199,6 +204,60 @@ post_config_update(?VALIDATIONS_CONF_PATH, {reorder, _Order}, New, _Old, _AppEnv
...
+    #{added := NewRawValidations} = emqx_utils:diff_lists(
+        ImportedRawValidations,
+        OldRawValidations,
+        fun(#{<<"name">> := N}) -> N end
+    ),
+    OtherConfs = maps:to_list(RawConf1),
+    Result = emqx_utils:foldl_while(
+        fun
+            ({validation, RawValidation}, Acc) ->
+                case insert(RawValidation) of
```
Comment (zmstone) [inline apps/emqx_schema_validation/src/emqx_schema_validation.erl:231]: this generates small config changes,
would it make sense to merge all configs and reload all at once?

**Review by zmstone — COMMENTED** [review]
i wonder if there are some inconsistent behaviour for config import comparing to config load API (which has a `mode` option supporting `replace` or `merge`)

**Comment by zmstone** [comment]
> i wonder if there are some inconsistent behaviour for config import comparing to config load API (which has a `mode` option supporting `replace` or `merge`)

There is maybe a missing support for config load API (and CLI) ?

File: apps/emqx_schema_validation/src/emqx_schema_validation.erl (hunk 1 above)
Comment (thalesmg) [reply apps/emqx_schema_validation/src/emqx_schema_validation.erl]: :heavy_check_mark:

File: apps/emqx_schema_validation/src/emqx_schema_validation.erl [hunk 2]
```diff
@@ -197,7 +209,75 @@ post_config_update(?VALIDATIONS_CONF_PATH, {delete, Name}, _New, Old, _AppEnvs)
     ok;
 post_config_update(?VALIDATIONS_CONF_PATH, {reorder, _Order}, New, _Old, _AppEnvs) ->
     ok = emqx_schema_validation_registry:reindex_positions(New),
-    ok.
+    ok;
+post_config_update([?CONF_ROOT], {merge, Input}, ResultingConfig, Old, _AppEnvs) ->
+    #{validations := ResultingValidations} = ResultingConfig,
+    #{new_validations := NewValidations0} = prepare_config_merge(Input, Old),
```
Comment (zhongwencool) [inline apps/emqx_schema_validation/src/emqx_schema_validation.erl:215]: The `Old` map of `prepare_config_merge/2` has atom as key in pre_config_update, but in post_config_update it is binary. This was not considered in the implementation.
By the way, why is there a need to do merge again in post after it has been done in pre_config_update? 
It seems to be some kind of conflict. Maybe, we could handle nothing in pre_config_update and deal with it directly in post_config_update.

Comment (thalesmg) [reply apps/emqx_schema_validation/src/emqx_schema_validation.erl:215]: > The Old map of prepare_config_merge/2 has atom as key in pre_config_update, but in post_config_update it is binary. This was not considered in the implementation.

Good catch.

> By the way, why is there a need to do merge again in post after it has been done in pre_config_update?
It seems to be some kind of conflict. Maybe, we could handle nothing in pre_config_update and deal with it directly in post_config_update.

Indeed, there's no work needed in pre_config_update.  Thanks!  :beers:

**Review by zhongwencool — APPROVED** [review]
LGTM, I merge this first, Let QA team help to verify this.

**Comment by emqxqa** [comment]
[TestExecution]()

**Pull request merged**